### PR TITLE
Static import support added.

### PR DESCRIPTION
**Pull request by sormuras**
Solves #73 like @swankjesse wants it. :sweat_smile:

Also includes a fix for a stack overflow that happend inside/between `hashCode()`s of `ParameterizedTypeName` and `TypeVariableName`.

### CODE REVIEW

File: src/main/java/com/squareup/javapoet/CodeWriter.java [hunk 1]
```diff
@@ -63,17 +68,25 @@
...
   }
 
-  CodeWriter(Appendable out, String indent, Map<String, ClassName> importedTypes) {
+  CodeWriter(Appendable out, String indent, Map<String, ClassName> importedTypes,
+      List<StaticImportItem> staticImportItems) {
     this.out = checkNotNull(out, "out == null");
     this.indent = checkNotNull(indent, "indent == null");
     this.importedTypes = checkNotNull(importedTypes, "importedTypes == null");
+    this.staticImportClassNames = new HashSet<>();
+    this.staticImportSignatures = new HashSet<>();
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: `LinkedHashSet` for both

File: src/main/java/com/squareup/javapoet/CodeWriter.java [hunk 2]
```diff
@@ -63,17 +68,25 @@
...
+    checkNotNull(staticImportItems, "staticImportItems == null");
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: Move up with other argument validation

File: src/main/java/com/squareup/javapoet/TypeName.java [hunk 3]
```diff
@@ -163,11 +163,12 @@ public TypeName unbox() {
     if (this == o) return true;
     if (o == null) return false;
     if (getClass() != o.getClass()) return false;
-    return toString().equals(o.toString());
+    return keyword.equals(((TypeName) o).keyword);
   }
 
   @Override public int hashCode() {
-    return toString().hashCode();
+    if (keyword == null) throw new AssertionError();
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/TypeName.java]: This seems really presumptuous. What if consumers are using a map or set for these types?

Comment (sormuras) [inline src/main/java/com/squareup/javapoet/TypeName.java]: See `emit(CodeWriter)` below, just mimicked that behaviour. Like `emit`, all subclasses of `TypeName` override `hashCode`, should work well with collections.

File: src/main/java/com/squareup/javapoet/CodeWriter.java (hunk 1 above)
Comment (sormuras) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: Sure.

File: src/main/java/com/squareup/javapoet/CodeWriter.java (hunk 2 above)
Comment (sormuras) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: Moved.

File: src/main/java/com/squareup/javapoet/TypeName.java (hunk 3 above)
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/TypeName.java]: Got it.

File: src/main/java/com/squareup/javapoet/TypeVariableName.java [hunk 4]
```diff
@@ -43,6 +44,7 @@ private TypeVariableName(String name, List<TypeName> bounds, List<AnnotationSpec
     super(annotations);
     this.name = checkNotNull(name, "name == null");
     this.bounds = bounds;
+    this.boundsHashCode = bounds.hashCode();
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/TypeVariableName.java]: is this problematic for cycle detection?

File: src/main/java/com/squareup/javapoet/CodeWriter.java [hunk 5]
```diff
@@ -247,6 +265,27 @@ public CodeWriter emit(CodeBlock codeBlock) throws IOException {
           break;
 
         default:
+          if (potentiallyStaticImported == null) {
+            emitAndIndent(part);
+            break;
+          }
+          ClassName pendingClassName = potentiallyStaticImported;
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: I think you want a helper method here. That way you can avoid the duplication between 268 and 288 ?

File: src/main/java/com/squareup/javapoet/CodeWriter.java [hunk 6]
```diff
@@ -247,6 +265,27 @@ public CodeWriter emit(CodeBlock codeBlock) throws IOException {
...
+          potentiallyStaticImported = null;
+          if (part.startsWith(".")) {
+              int endIndex = part.indexOf('(');
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: or any non-word character, right? like `System.out`

Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: (The part might be `.out.println()`)

File: src/main/java/com/squareup/javapoet/CodeWriter.java [hunk 7]
```diff
@@ -247,6 +265,27 @@ public CodeWriter emit(CodeBlock codeBlock) throws IOException {
...
+              if (endIndex < 0) endIndex = part.length();
+              String memberName = part.substring(1, endIndex);
+              checkState(isIdentifier(memberName), "identifier expected, but got: %s", memberName);
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: I’m guessing this explodes on `System.out.println()`

File: src/main/java/com/squareup/javapoet/CodeWriter.java [hunk 8]
```diff
@@ -247,6 +265,27 @@ public CodeWriter emit(CodeBlock codeBlock) throws IOException {
...
+              String candidate = pendingClassName.canonicalName + "." + memberName;
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: Neat

File: src/main/java/com/squareup/javapoet/CodeWriter.java [hunk 9]
```diff
@@ -247,6 +265,27 @@ public CodeWriter emit(CodeBlock codeBlock) throws IOException {
...
+              String wildcard = pendingClassName.canonicalName + ".*";
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: Neat.

File: src/main/java/com/squareup/javapoet/CodeWriter.java [hunk 10]
```diff
@@ -247,6 +265,27 @@ public CodeWriter emit(CodeBlock codeBlock) throws IOException {
...
+              if (staticImportSignatures.contains(candidate)
+                  || staticImportSignatures.contains(wildcard)) {
+                String partTail = part.substring(memberName.length() + 1);
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: Instead of this line and 284, isn't it just, `emitAndIndent(part.substring(1));` ?

Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: Ie. chomp the dot ?

File: src/main/java/com/squareup/javapoet/CodeWriter.java [hunk 11]
```diff
@@ -247,6 +265,27 @@ public CodeWriter emit(CodeBlock codeBlock) throws IOException {
...
+                emitAndIndent(memberName + partTail);
+                break;
+              }
+          }
+          pendingClassName.emit(this);
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: Neat.

File: src/main/java/com/squareup/javapoet/JavaFile.java [hunk 12]
```diff
@@ -40,6 +42,22 @@
 
 /** A Java file containing a single top level class. */
 public final class JavaFile {
+  static final class StaticImportItem implements Comparable<StaticImportItem> {
+    final ClassName className;
+    final String memberName;
+    final String signature;
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/JavaFile.java]: Do we need anything other than signature?

**Comment by swankjesse** [comment]
I like where this is going.

File: src/main/java/com/squareup/javapoet/TypeVariableName.java (hunk 4 above)
Comment (sormuras) [inline src/main/java/com/squareup/javapoet/TypeVariableName.java]: Both, `TypesTest` and `TypesEclipseTest`, run into a cycle, when line 117 in `AbstractTypesTest.getTypeVariableTypeMirrorRecursive()` the statement `typeVariableName.bounds.toString()` is executed. But only when:
 * `ParameterizedTypeName.hashCode()` and `TypeVariableName.hashCode()` calculate the internal collection hash code *on the fly*
 * it only happens if run via Maven/javac - if started inside Eclipse, there's no stack overflow produced

See failure report:
```
	at com.squareup.javapoet.ParameterizedTypeName.hashCode(ParameterizedTypeName.java:64)
	at java.util.AbstractList.hashCode(AbstractList.java:541)
	at java.util.Collections$UnmodifiableList.hashCode(Collections.java:1307)
	at com.squareup.javapoet.TypeVariableName.hashCode(TypeVariableName.java:72)
	at java.util.AbstractList.hashCode(AbstractList.java:541)
	at java.util.Collections$UnmodifiableList.hashCode(Collections.java:1307)
	at com.squareup.javapoet.ParameterizedTypeName.hashCode(ParameterizedTypeName.java:64)
	at java.util.AbstractList.hashCode(AbstractList.java:541)
	at java.util.Collections$UnmodifiableList.hashCode(Collections.java:1307)
	at com.squareup.javapoet.ParameterizedTypeName.hashCode(ParameterizedTypeName.java:64)


Results :

Tests in error: 
  getTypeVariableTypeMirrorRecursive(com.squareup.javapoet.TypesEclipseTest)
  getTypeVariableTypeMirrorRecursive(com.squareup.javapoet.TypesTest)

Tests run: 237, Failures: 0, Errors: 2, Skipped: 0
```
I move the entire hash code calculation into the constructor. Both classes are immutable and marked as final.

File: src/main/java/com/squareup/javapoet/CodeWriter.java (hunk 7 above)
Comment (sormuras) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: **BäMM!** It exploded and was removed.

File: src/main/java/com/squareup/javapoet/JavaFile.java (hunk 12 above)
Comment (sormuras) [inline src/main/java/com/squareup/javapoet/JavaFile.java]: Needed the canonicle class name for indentification of potentially imported `$T`s ... will try to boil it down to the signature string, as the `if (staticImportClassNames.contains(typeName))` condition in `case "$T":` could be written as `if (staticImportSignatures.contains(typeName.toString()))`

File: src/main/java/com/squareup/javapoet/CodeWriter.java (hunk 10 above)
Comment (sormuras) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: True.

File: src/main/java/com/squareup/javapoet/CodeWriter.java (hunk 5 above)
Comment (sormuras) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: I remove the first block, as the logic can be moved to the second.

File: src/main/java/com/squareup/javapoet/JavaFile.java (hunk 12 above)
Comment (sormuras) [inline src/main/java/com/squareup/javapoet/JavaFile.java]: Worked out.

**Comment by sormuras** [comment]
Some local tests with mixed normal/static import fail. Fixing...

**Comment by sormuras** [comment]
Tests and fixes committed. `SourceVersion.isIdentifier()` is back for good.

File: src/main/java/com/squareup/javapoet/CodeWriter.java [hunk 13]
```diff
@@ -218,6 +227,14 @@ public CodeWriter emit(CodeBlock codeBlock) throws IOException {
 
         case "$T":
           TypeName typeName = (TypeName) codeBlock.args.get(a++);
+          if (typeName instanceof ClassName) {
+            ClassName candidate = (ClassName) typeName;
+            if (staticImportClassNames.contains(candidate.canonicalName)) {
+              checkState(potentiallyStaticImported == null, "pending type for static import?!");
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: ooh interesting. Like in say, `$T$T`.

File: src/main/java/com/squareup/javapoet/JavaFile.java [hunk 14]
```diff
@@ -131,6 +135,13 @@ private void emit(CodeWriter codeWriter) throws IOException {
       codeWriter.emit("\n");
     }
 
+    if (!staticImports.isEmpty()) {
+      for (String signature : staticImports) {
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/JavaFile.java]: sort these?

Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/JavaFile.java]: or maybe just use a `TreeSet` for ’em? Order is never important, and that saves a List to Set conversion later!

File: src/main/java/com/squareup/javapoet/JavaFile.java [hunk 15]
```diff
@@ -218,6 +230,25 @@ public Builder addFileComment(String format, Object... args) {
...
 
+    public Builder addStaticImport(Enum<?> constant) {
+      return addStaticImport(ClassName.get(constant.getDeclaringClass()), constant.name());
+    }
+
+    public Builder addStaticImport(Class<?> clazz, String... names) {
+      return addStaticImport(ClassName.get(clazz), names);
+    }
+
+    public Builder addStaticImport(ClassName className, String... names) {
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/JavaFile.java]: nice.

File: src/main/java/com/squareup/javapoet/JavaFile.java [hunk 16]
```diff
@@ -218,6 +230,25 @@ public Builder addFileComment(String format, Object... args) {
       return this;
     }
 
+    public Builder addStaticImport(Enum<?> constant) {
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/JavaFile.java]: neat.

File: src/main/java/com/squareup/javapoet/JavaFile.java [hunk 17]
```diff
@@ -237,6 +268,7 @@ public Builder indent(String indent) {
     }
 
     public JavaFile build() {
+      Collections.sort(staticImports);
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/JavaFile.java]: ah, there it is.

Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/JavaFile.java]: I think it’s slightly neater to keep it in a sorted set, but it doesn’t really matter!

Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/JavaFile.java]: (I think my reluctance about this is that `build()` is mutating. A crazy person might call `build()` from two threads in parallel, and crash.)

File: src/main/java/com/squareup/javapoet/Util.java [hunk 18]
```diff
@@ -109,4 +110,19 @@ public static void requireExactlyOneOf(Set<Modifier> modifiers, Modifier... mutu
   public static boolean hasDefaultModifier(Collection<Modifier> modifiers) {
     return DEFAULT != null && modifiers.contains(DEFAULT);
   }
+
+  public static String identifier(String part) {
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/Util.java]: Maybe rename to identifierPrefix? Should this Just Work for any string, and return null if it doesn't start with an identifier character? That might make the calling code simpler.

File: src/main/java/com/squareup/javapoet/CodeWriter.java [hunk 19]
```diff
@@ -247,6 +264,19 @@ public CodeWriter emit(CodeBlock codeBlock) throws IOException {
           break;
 
         default:
+          ClassName pendingClassName = potentiallyStaticImported;
+          potentiallyStaticImported = null;
+          if (pendingClassName != null && part.startsWith(".")) {
+              String partWithoutLeadingDot = part.substring(1);
+              String memberName = Util.identifier(partWithoutLeadingDot);
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: I think this might crash on input like `$T...` ?

Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: because `Util.identifier` assumes partWithoutLeadingDot is an identifer.

File: src/main/java/com/squareup/javapoet/Util.java [hunk 20]
```diff
@@ -109,4 +110,19 @@ public static void requireExactlyOneOf(Set<Modifier> modifiers, Modifier... mutu
...
+    checkArgument(part != null, "part == null");
+    checkArgument(!part.isEmpty(), "part is empty");
+    checkArgument(Character.isJavaIdentifierStart(part.charAt(0)), "part[0] is not valid");
+    StringBuilder builder = new StringBuilder();
+    for (int i = 0; i < part.length(); i++) {
+      builder.append(part.charAt(i));
+      if (!SourceVersion.isIdentifier(builder.toString())) {
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/Util.java]: Probably not necessary to keep a string builder, you can just do successive calls to `part.substring(0, i)`. With the builder you make more work for yourself by having to pop.

File: src/test/java/com/squareup/javapoet/JavaFileTest.java [hunk 21]
```diff
@@ -25,6 +26,119 @@
...
+        + "\n"
+        + "class Util {\n"
+        + "  public static long minutesToSeconds(long minutes) {\n"
+        + "    System.gc();\n"
+        + "    return SECONDS.convert(minutes, TimeUnit.MINUTES);\n"
+        + "  }\n"
+        + "}\n");
+  }
+
+  @Test public void importStaticTwice() {
```
Comment (swankjesse) [inline src/test/java/com/squareup/javapoet/JavaFileTest.java]: These tests are perfect.

File: src/test/java/com/squareup/javapoet/JavaFileTest.java [hunk 22]
```diff
@@ -25,6 +26,119 @@
...
+            + "\n"
+            + "class Util {\n"
+            + "  public static long minutesToSeconds(long minutes) {\n"
+            + "    System.gc();\n"
+            + "    return SECONDS.convert(minutes, MINUTES);\n"
+            + "  }\n"
+            + "}\n");
+  }
+
+  @Test public void importStaticTwiceAndGC() {
```
Comment (swankjesse) [inline src/test/java/com/squareup/javapoet/JavaFileTest.java]: The GC is a weird way to name this!

**Comment by swankjesse** [comment]
I like where this is going. I think I found one little bug, otherwise it's perfect. YES!

File: src/main/java/com/squareup/javapoet/CodeWriter.java (hunk 13 above)
Comment (sormuras) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: Ja. Needs a test case to show...

File: src/main/java/com/squareup/javapoet/JavaFile.java (hunk 14 above)
Comment (sormuras) [inline src/main/java/com/squareup/javapoet/JavaFile.java]: The list is a left-over from the `StaticImportItem` class. Will convert it to a sorted set.

File: src/main/java/com/squareup/javapoet/CodeWriter.java (hunk 19 above)
Comment (sormuras) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: Good catch. My mind is still using the `$R` syntax, in which these edge cases are not present.

File: src/test/java/com/squareup/javapoet/JavaFileTest.java (hunk 22 above)
Comment (sormuras) [inline src/test/java/com/squareup/javapoet/JavaFileTest.java]: Rename inc.

File: src/main/java/com/squareup/javapoet/Util.java (hunk 18 above)
Comment (sormuras) [inline src/main/java/com/squareup/javapoet/Util.java]: The only usage is quite simple. :smile: Only put the method into `Util` because of it's easier to test. Move back to `JavaFile` and make it private? It's not intended to be a general purpose method.

Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/Util.java]: Yeah, I don’t like special-purpose methods in general-purpose locations.

Comment (sormuras) [inline src/main/java/com/squareup/javapoet/Util.java]: Alright. Will add tests that show the intended (and invalid) input strings. Your `$T...` comment does not apply, as `$T` is already resolved and every varargs method emitted didn't throw an exception for the entire test suite. You can't statically import a sole type name, can you?

Comment (sormuras) [inline src/main/java/com/squareup/javapoet/Util.java]: I mean, if s/o produces syntax errors ... the compiler will tell anyway.
* `$T$T` will always crash the compiler
* `$T...` can't be used to declare the last method parameter in JavaPoet-style? And if it's found inside a generic `CodeBlock` ... well.

Comment (sormuras) [inline src/main/java/com/squareup/javapoet/Util.java]: I revoke my last two comments.

            .addMethod(MethodSpec.constructorBuilder()
                .addParameter(Thread.State[].class, "states")
                .varargs(true)
                .build())

produces your prediction

	java.lang.IllegalArgumentException: part[0] is not valid
		at com.squareup.javapoet.Util.checkArgument(Util.java:63)
		at com.squareup.javapoet.Util.identifier(Util.java:117)
		at com.squareup.javapoet.CodeWriter.emit(CodeWriter.java:271)

File: src/main/java/com/squareup/javapoet/JavaFile.java (hunk 16 above)
Comment (sormuras) [inline src/main/java/com/squareup/javapoet/JavaFile.java]: I'm getting closer and closer... it's been [$E](https://github.com/sormuras/javapoet/commit/ba36bff2656b7215d440239b0c73065a3a0cfdfd), it's still [$R](https://github.com/square/javapoet/pull/388) ... for now, I can live with **$T**, when it comes to static imports.

File: src/test/java/com/squareup/javapoet/JavaFileTest.java [hunk 23]
```diff
@@ -25,6 +26,141 @@
 
 @RunWith(JUnit4.class)
 public final class JavaFileTest {
+  @Test(expected = IllegalStateException.class)
+  public void importStaticFailsForChainedTypeName() {
```
Comment (sormuras) [inline src/test/java/com/squareup/javapoet/JavaFileTest.java]: Keep this state check "as is", or should I work around this strange case of `TypeName` chaining? Maybe like, emit the old pendine one and remember the current for the next round?

File: src/main/java/com/squareup/javapoet/CodeWriter.java (hunk 13 above)
Comment (sormuras) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: It actually throws for anything that looks like: `$T*$T` with `*` is any number of non-default parts:

    addStatement("$1T$2L$1T", Runtime.class, "?") // illegal if $T is static imported

Comment (sormuras) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: I'll fix/guard that by peeking into the next part and check whether `!part.startsWith("$")` -- ensuring that only if a `$T` is followed by a default case, the emittion of `$T` is defered.

**Comment by sormuras** [comment]
Feels like the final iteration (including a ListIterator) now.

File: src/main/java/com/squareup/javapoet/CodeWriter.java [hunk 24]
```diff
@@ -254,6 +290,29 @@ public CodeWriter emit(CodeBlock codeBlock) throws IOException {
     return this;
   }
 
+  private static String gleanMemberName(String part) {
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: I might have to rename this later to avoid the word glean.

File: src/main/java/com/squareup/javapoet/CodeWriter.java [hunk 25]
```diff
@@ -254,6 +290,29 @@ public CodeWriter emit(CodeBlock codeBlock) throws IOException {
...
+      if (!SourceVersion.isIdentifier(part.substring(0, i))) {
+        return part.substring(0, i - 1);
+      }
+    }
+    return part;
+  }
+
+  private boolean emitStaticImportMember(String canonical, String part) throws IOException {
+    String partWithoutLeadingDot = part.substring(1);
+    char first = partWithoutLeadingDot.charAt(0);
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: Does this explode if the part is a single character?

File: src/main/java/com/squareup/javapoet/WildcardTypeName.java [hunk 26]
```diff
@@ -42,6 +43,7 @@ private WildcardTypeName(List<TypeName> upperBounds, List<TypeName> lowerBounds,
     super(annotations);
     this.upperBounds = Util.immutableList(upperBounds);
     this.lowerBounds = Util.immutableList(lowerBounds);
+    this.hashCode = upperBounds.hashCode() ^ lowerBounds.hashCode();
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/WildcardTypeName.java]: I’m assuming theses change are unrelated, right?

**Comment by swankjesse** [comment]
I suspect there’s a few unlikely corner cases, such as when the _last_ format part is a `$T`, but this is certainly better than what we have now. I might take a pass at this code to get rid of the nasty word _glean_ and to try to tighten up those exotic & unlikely cases. I’m also tempted to try to make this work with `$L`, so you don’t need to statically know what you’re statically importing!

**Comment by swankjesse** [comment]
And `$N`.

File: src/main/java/com/squareup/javapoet/CodeWriter.java (hunk 24 above)
Comment (sormuras) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: Preparing a *static_polish* PR, with `extractMemberName`

File: src/main/java/com/squareup/javapoet/CodeWriter.java (hunk 25 above)
Comment (sormuras) [inline src/main/java/com/squareup/javapoet/CodeWriter.java]: Dooh. Late time optimization, the `.isEmpty()` check is too late now. Gonna fix it with a test case attached.

File: src/main/java/com/squareup/javapoet/WildcardTypeName.java (hunk 26 above)
Comment (sormuras) [inline src/main/java/com/squareup/javapoet/WildcardTypeName.java]: Yes and no. The precalc of hash code in `ParameterizedTypeName` is neccessary - the precalcs in `TypeVariableName` and `WildcardTypeName` are here, to a) make it look and work similar and b) prevent future stack overflows.
